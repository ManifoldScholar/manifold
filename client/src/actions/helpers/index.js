export { default as createApiAction } from "./createApiAction";
export { default as commonActions } from "./commonActions";
