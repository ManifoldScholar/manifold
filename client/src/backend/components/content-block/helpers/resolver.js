import blocks from "../Builder/Block/types";
import forms from "../TypeForm/types";
import contentBlockHelpers from "helpers/contentBlockHelpers";
import transform from "lodash/transform";

class Resolver {
  blockComponentsByType() {
    /* eslint-disable no-param-reassign */
    return transform(blocks, (result, val, key) => {
      result[contentBlockHelpers.componentKeyToType(key)] = val;
    });
    /* eslint-enable no-param-reassign */
  }

  typeToBlockComponent(type) {
    return contentBlockHelpers.typeToComponent(type, blocks);
  }

  typeToFormComponent(type) {
    return contentBlockHelpers.typeToComponent(type, forms);
  }
}

export default new Resolver();
