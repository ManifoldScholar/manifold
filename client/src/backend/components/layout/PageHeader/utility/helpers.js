import lh from "helpers/linkHandler";

export const getTextLinks = ({ texts, pathname }) =>
  texts?.length
    ? texts?.map(t => ({
        label: t.label,
        route: "backendTextAnalytics",
        id: t.id,
        active: pathname?.includes(t.id)
      }))
    : null;

export const getIssueLinks = ({ issues, pathname }) =>
  issues?.length
    ? issues?.map(i => ({
        label: i.label,
        route: "backendProjectAnalytics",
        id: i.id,
        active: pathname?.includes(i.id)
      }))
    : null;

export const getLinkOrButtonProps = action => {
  if (action.href) {
    return action.download
      ? { as: "a", href: action.href, download: true }
      : {
          as: "a",
          href: action.href,
          target: "_blank",
          re: "noopener noreferrer"
        };
  }

  if (action.route)
    return {
      to: lh.link(action.route, action.slug, action.resourceSlug)
    };

  return { as: "button", onClick: action.onClick };
};
