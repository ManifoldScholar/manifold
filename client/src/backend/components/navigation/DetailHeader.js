import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import Dropdown from "./Dropdown";
import { Link } from "react-router-dom";
import classNames from "classnames";
import IconComposer from "global/components/utility/IconComposer";
import UniqueIcons from "global/components/icon/unique";

export default class DetailHeader extends PureComponent {
  static displayName = "Navigation.DetailHeader";

  static propTypes = {
    type: PropTypes.string.isRequired,
    iconName: PropTypes.string,
    title: PropTypes.string.isRequired,
    subtitle: PropTypes.string,
    utility: PropTypes.object,
    note: PropTypes.string,
    secondaryLinks: PropTypes.array,
    backUrl: PropTypes.string,
    backLabel: PropTypes.string
  };

  get type() {
    return this.props.type;
  }

  get isProject() {
    return this.type === "project";
  }

  get isProjectCollection() {
    return this.type === "projectCollection";
  }

  get title() {
    return this.props.title;
  }

  get subtitle() {
    return this.props.subtitle;
  }

  get utility() {
    return this.props.utility || null;
  }

  get note() {
    return this.props.note;
  }

  get backLabel() {
    return this.props.backLabel || "Back";
  }

  get backUrl() {
    return this.props.backUrl;
  }

  get backHiddenDesktop() {
    return this.isProjectCollection;
  }

  get secondaryLinks() {
    return this.props.secondaryLinks;
  }

  get figureClasses() {
    return classNames("backend-header__figure", {
      "backend-header__figure--rounded": this.isProjectCollection,
      "backend-header__figure--accented": this.isProjectCollection
    });
  }

  get iconName() {
    return this.props.iconName;
  }

  get iconFromType() {
    if (this.iconName) return this.iconName;

    let icon = "";
    switch (this.type) {
      case "feature":
        icon = "Lamp64";
        break;
      case "page":
        icon = "ResourceDocument64";
        break;
      case "resourceCollection":
        icon = "ResourceCollection64";
        break;
      case "resources":
        icon = "BEResourcesBoxes64";
        break;
      case "resource":
        icon = "BEResourcesBox64";
        break;
      case "text":
        icon = "TextsStacked64";
        break;
      default:
    }

    return icon;
  }

  projectIcon() {
    return (
      <UniqueIcons.ProjectPlaceholderUnique
        mode="small"
        color="outlined"
        iconClass="backend-header__type-icon backend-header__type-icon--project"
      />
    );
  }

  typeIcon() {
    return (
      <IconComposer
        icon={this.iconFromType}
        size="default"
        iconClass="backend-header__type-icon"
      />
    );
  }

  breadcrumbs() {
    return (
      <nav
        className={classNames("backend-header__breadcrumbs", "breadcrumbs", {
          "breadcrumbs--hidden-desktop": this.backHiddenDesktop
        })}
      >
        <div className="breadcrumbs__inner">
          <Link to={this.backUrl} className="breadcrumbs__link">
            <IconComposer
              icon="arrowLeft16"
              size="default"
              iconClass="breadcrumbs__icon breadcrumbs__icon--small"
            />
            <IconComposer
              icon="arrowLeft32"
              size="default"
              iconClass="breadcrumbs__icon breadcrumbs__icon--large"
            />
            {this.backLabel}
          </Link>
        </div>
      </nav>
    );
  }

  sectionNav() {
    return (
      <Dropdown
        classNames="backend-header__section-nav"
        links={this.secondaryLinks}
      />
    );
  }

  figureBlock() {
    return (
      <figure className="backend-header__figure-block">
        <div className={this.figureClasses}>
          {this.isProject ? this.projectIcon() : this.typeIcon()}
        </div>
      </figure>
    );
  }

  titleBlock() {
    return (
      <div className="backend-header__title-block">
        <h1 className="backend-header__title">
          <span dangerouslySetInnerHTML={{ __html: this.title }} />
          <span className="backend-header__subtitle">{this.subtitle}</span>
        </h1>
      </div>
    );
  }

  utilityBlock() {
    return (
      <div className="backend-header__utility-block">
        {this.utility}
        {this.note && <span className="backend-header__note">{this.note}</span>}
      </div>
    );
  }

  render() {
    return (
      <section className="backend-header">
        {this.backUrl && this.breadcrumbs()}
        <div className="backend-header__inner backend-header__inner--padded">
          <header className="backend-header__content-wrapper">
            {this.titleBlock()}
            {this.utilityBlock()}
            {this.figureBlock()}
          </header>
        </div>
        {this.secondaryLinks && this.sectionNav()}
      </section>
    );
  }
}
