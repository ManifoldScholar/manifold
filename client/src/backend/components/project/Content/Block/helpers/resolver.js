import types from "../types";

class Resolver {
  typeToComponentKey(type) {
    const matches = type.match(/^Content::(.*)Block$/);
    if (!matches) return null;
    return matches[1];
  }

  typeToComponent(type) {
    const key = this.typeToComponentKey(type);
    if (!key) return null;
    const component = types[key];
    return component || null;
  }

  typeToAvailableId(type) {
    return type;
  }

  typeToCurrentId(id) {
    return `current-${id}`;
  }
}

export default new Resolver();
