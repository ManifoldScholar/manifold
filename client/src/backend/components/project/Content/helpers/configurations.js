class Configurations {
  get configurationMap() {
    return {
      "Content::HeroBlock": true,
      "Content::RecentActivityBlock": false,
      "Content::TextsBlock": true,
      "Content::ResourcesBlock": true,
      "Content::MetadataBlock": false,
      "Content::TableOfContentsBlock": true,
      "Content::MarkdownBlock": true
    };
  }

  isConfigurable(type) {
    return this.configurationMap[type];
  }
}

export default new Configurations();
