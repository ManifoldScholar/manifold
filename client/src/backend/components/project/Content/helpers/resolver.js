import blocks from "../Block/types";
import forms from "../TypeForm/types";

class Resolver {
  typeToComponentKey(type) {
    const matches = type.match(/^Content::(.*)Block$/);
    if (!matches) return null;
    return matches[1];
  }

  typeToComponent(type, types) {
    const key = this.typeToComponentKey(type);
    if (!key) return null;
    const component = types[key];
    return component || null;
  }

  typeToBlockComponent(type) {
    return this.typeToComponent(type, blocks);
  }

  typeToFormComponent(type) {
    return this.typeToComponent(type, forms);
  }

  typeToAvailableId(type) {
    return type;
  }

  typeToCurrentId(id) {
    return `current-${id}`;
  }
}

export default new Resolver();
