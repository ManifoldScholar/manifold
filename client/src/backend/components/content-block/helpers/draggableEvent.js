import startsWith from "lodash/startsWith";
import resolver from "./resolver";
import sumBy from "lodash/sumBy";
import cloneDeep from "lodash/cloneDeep";

export default class DraggableEventHelper {
  static syntheticDraggable = (blockType, opts = {}) => {
    const options = {
      ...DraggableEventHelper.defaultOptions(),
      ...opts
    };
    return {
      type: DraggableEventHelper.isTopType(blockType) ? "TOP" : "BOTTOM",
      draggableId: blockType,
      source: {
        droppableId: options.source.name
      },
      destination: {
        droppableId: options.destination.name,
        index: -1
      }
    };
  };

  static isTopType = type => {
    return !!resolver.typeToBlockComponent(type).top;
  };

  static defaultOptions = () => {
    return {
      source: {
        name: "available"
      },
      destination: {
        name: "current"
      }
    };
  };

  constructor(draggable, blocks, options = {}) {
    this.options = { ...this.defaultOptions, ...options };
    this.draggable = draggable;
    this._blocks = blocks;
  }

  get block() {
    if (this.isMove) return this.movedBlock;
    if (this.isInsert) return this.insertedBlock;
  }

  get blocks() {
    let blocks = this.clonedBlocks;
    const block = this.block;
    if (this.isMove) {
      blocks = blocks.filter(b => b.id !== block.id);
    }
    blocks.splice(block.attributes.position - 1, 0, block);
    return blocks;
  }

  get defaultOptions() {
    return this.constructor.defaultOptions();
  }

  get clonedBlocks() {
    return Array.from(this._blocks);
  }

  get draggedFromSource() {
    if (!this.draggable.source) return false;
    return startsWith(
      this.draggable.source.droppableId,
      this.options.source.name
    );
  }

  get draggedFromDestination() {
    if (!this.draggable.source) return false;
    return startsWith(
      this.draggable.source.droppableId,
      this.options.destination.name
    );
  }

  get draggedToDestination() {
    if (!this.draggable.destination) return false;
    return startsWith(
      this.draggable.destination.droppableId,
      this.options.destination.name
    );
  }

  get destinationIndex() {
    if (!this.draggable.destination) return null;
    const index = this.draggable.destination.index;
    if (index === -1) return this.autoIndex;
    return index;
  }

  get autoIndex() {
    if (this.isTopType(this.type)) return this.topCount; // no +1 because index is from 0;
    return this._blocks.length;
  }

  get actionable() {
    return this.action != null;
  }

  get isMove() {
    return this.draggedFromDestination && this.draggedToDestination;
  }

  get isInsert() {
    return this.draggedFromSource && this.draggedToDestination;
  }

  get action() {
    if (this.isMove) return "move";
    if (this.isInsert) return "insert";
    return null;
  }

  get isTop() {
    return this.draggable.type === "TOP";
  }

  isTopBlock(block) {
    return this.isTopType(block.attributes.type);
  }

  isTopType(type) {
    return this.constructor.isTopType(type);
  }

  get topCount() {
    return sumBy(this._blocks, block => (this.isTopBlock(block) ? 1 : 0));
  }

  get type() {
    return this.draggable.draggableId;
  }

  get uuid() {
    return this.draggable.draggableId;
  }

  get position() {
    const index = this.destinationIndex;
    const offset = 1; // position starts from 1, index from 0
    if (this.isTop) return index + offset;
    // Blocks in the bottom list have to add the top count.
    return index + this.topCount + offset;
  }

  findBlock(id) {
    return this._blocks.find(block => block.id === id);
  }

  cloneBlock(id) {
    return cloneDeep(this.findBlock(id));
  }

  get insertedBlock() {
    return {
      id: "pending",
      attributes: {
        type: this.type,
        position: this.position,
        renderable: true
      },
      relationships: {}
    };
  }

  get movedBlock() {
    const block = this.cloneBlock(this.uuid);
    block.attributes.position = this.position;
    return block;
  }
}
