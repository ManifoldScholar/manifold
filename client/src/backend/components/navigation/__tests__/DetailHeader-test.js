import React from "react";
import renderer from "react-test-renderer";
import DetailHeader from "../DetailHeader";
import { wrapWithRouter, renderWithRouter } from "test/helpers/routing";

describe("Backend.Navigation.DetailHeader component", () => {
  it("renders correctly", () => {
    const component = renderer.create(
      wrapWithRouter(
        <DetailHeader
          title="Rowan"
          subtitle="World's Greatest Dog"
          type="project"
          breadcrumb={[
            {
              label: "All Projects",
              path: "/backend"
            }
          ]}
        />
      )
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
