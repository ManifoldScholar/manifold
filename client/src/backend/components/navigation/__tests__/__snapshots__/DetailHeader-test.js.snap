// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backend.Navigation.DetailHeader component renders correctly 1`] = `
<section
  className="backend-header"
>
  <div
    className="backend-header__inner backend-header__inner--padded"
  >
    <header
      className="backend-header__content-wrapper"
    >
      <div
        className="backend-header__title-block"
      >
        <h1
          className="backend-header__title"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "Rowan",
              }
            }
          />
          <span
            className="backend-header__subtitle"
          >
            World's Greatest Dog
          </span>
        </h1>
      </div>
      <div
        className="backend-header__utility-block"
      />
      <figure
        className="backend-header__figure-block"
      >
        <div
          className="backend-header__figure"
        >
          <svg
            aria-label="Default Project Thumbnail"
            className="project-thumb-placeholder backend-header__type-icon backend-header__type-icon--project project-thumb-placeholder--all project-thumb-placeholder--outlined"
            height="100%"
            viewBox="0 0 48 48"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fillRule="evenodd"
            >
              <polygon
                className="project-thumb-placeholder__tile"
                fill="#CBF7E6"
                points="0 42 42 42 42 0 0 0"
              />
              <g
                className="project-thumb-placeholder__illustration"
                transform="translate(9 8)"
              >
                <g
                  transform="translate(20.683 2.244)"
                >
                  <polygon
                    points=".161 2.954 .161 21.013 1.612 23.598 3.06 21.013 3.06 2.954"
                  />
                  <path
                    d="M3.06081501,21.012797 L0.161633021,21.012797"
                  />
                  <path
                    d="M0.162011257,2.95560225 L3.06119325,2.95560225 L3.06119325,1.53299812 C3.06119325,0.980713374 2.613478,0.532998124 2.06119325,0.532998124 L1.16201126,0.532998124 C0.609726507,0.532998124 0.162011257,0.980713374 0.162011257,1.53299812 L0.162011257,2.95560225 Z"
                    strokeLinejoin="round"
                  />
                </g>
                <g
                  transform="translate(.195 11.463)"
                >
                  <polygon
                    points="14.898 .419 .985 .419 .985 11.282 8.141 11.282 8.141 14.378 10.844 11.282 14.898 11.282"
                  />
                  <path
                    d="M12.5237854 3.46905816L3.32696735 3.46905816M12.5237854 8.00165403L3.32696735 8.00165403M12.5237854 5.73507242L3.32696735 5.73507242"
                  />
                </g>
                <g
                  transform="translate(.195 .195)"
                >
                  <path
                    d="M11.8853223 5.12585966C11.8853223 6.3608015 10.8848871 7.36123677 9.64994522 7.36123677 8.41689456 7.36123677 7.41645929 6.3608015 7.41645929 5.12585966 7.41645929 3.89091782 8.41689456 2.89048255 9.64994522 2.89048255 10.8848871 2.89048255 11.8853223 3.89091782 11.8853223 5.12585966zM11.8853223 5.23573734L15.0417051 2.0793546C15.6109508 1.51010882 16.5849096 1.91293058 16.5849096 2.71857411M3.99814784 3.09718874L6.63067317.464663415C7.20181013-.106473546 8.17576886.2982394 8.17576886 1.10388293"
                  />
                  <path
                    d="M5.22155347 5.12585966C5.22155347 6.3608015 4.2211182 7.36123677 2.98617636 7.36123677 1.7531257 7.36123677.752690432 6.3608015.752690432 5.12585966.752690432 3.89091782 1.7531257 2.89048255 2.98617636 2.89048255 4.2211182 2.89048255 5.22155347 3.89091782 5.22155347 5.12585966zM5.22155347 5.34334559L5.71136961 5.13720675C6.09906191 4.97267392 6.53781614 4.97267392 6.92739962 5.13720675L7.41532458 5.34334559"
                  />
                </g>
              </g>
              <g
                className="project-thumb-placeholder__frame"
                stroke="#828282"
              >
                <polyline
                  points="48 6 48 48 6 48"
                />
                <polyline
                  points="45 3 45 45 3 45"
                />
                <polygon
                  points="0 42 42 42 42 0 0 0"
                />
              </g>
            </g>
          </svg>
        </div>
      </figure>
    </header>
  </div>
</section>
`;
